DAO/src/lib.rs

```diff
@@ -0,0 +1,698 @@
+            },
+            exec::gas_available() - GAS_RESERVE,
+            0,
+        );
+    }
+
+    // Cancels the proposal after the end of the voting period if there are no YES votes.
+    // Requirements:
+    // * `msg::source()` must be the proposer
+    // * It can be cancelled if the number of YES votes is less than number of NO votes or the required quorum is not achieved
+    // * The voting period must be over
+    // * The proposal must not be cancelled or aborted
+    // Arguments:
+    // * `proposal_id`: the proposal ID
+    async fn cancel_proposal(&mut self, proposal_id: u128) {
+        if !self.proposals.contains_key(&proposal_id) {
+            panic!("proposal does not exist");
+        }
+        if self.proposals.get(&proposal_id).unwrap().proposer != msg::source() {
+            panic!("caller must be proposer");
+        }
+        let proposal = self.proposals.get_mut(&proposal_id).unwrap();
+        if proposal.yes_votes > proposal.no_votes
+            && proposal.yes_votes * 1000 / self.total_shares > proposal.quorum
+        {
+            panic!(
+                "Proposal can not be cancelled since YES votes > NO votes and quorum is achieved"
+            );
+        }
+        if exec::block_timestamp() < proposal.starting_period + self.voting_period_length {
+            panic!("The voting period is not over yet");
+        }
+        if proposal.cancelled || proposal.aborted {
+            panic!("Proposal has already been cancelled or aborted");
+        }
+        let amount = proposal.token_tribute;
+        proposal.token_tribute = 0;
+        proposal.cancelled = true;
+
+        transfer_tokens(&self.approved_token_program_id, &proposal.applicant, amount).await;
+
+        msg::reply(
+            DaoEvent::Cancel {
+                member: msg::source(),
+                proposal_id,
+            },
+            exec::gas_available() - GAS_RESERVE,
+            0,
+        );
+    }
+
+    // Aborts the membership proposal. It can be used in case when applicant is disagree with the requested shares or the details the proposer  indicated by the proposer
+    // Requirements:
+    // * `msg::source()` must be the applicant
+    // * The proposal must be membership proposal
+    // * The proposal can be aborted during only the abort window
+    // * The proposal must not be aborted yet
+    // Arguments:
+    // * `proposal_id`: the proposal ID
+    async fn abort(&mut self, proposal_id: u128) {
+        if !self.proposals.contains_key(&proposal_id) {
+            panic!("proposal does not exist");
+        }
+        if self.proposals.get(&proposal_id).unwrap().applicant != msg::source() {
+            panic!("caller must be applicant");
+        }
+        let proposal = self.proposals.get_mut(&proposal_id).unwrap();
+        if exec::block_timestamp() > proposal.starting_period + self.abort_window {
+            panic!("The abort window is over");
+        }
+        if proposal.aborted {
```

Comment:
suggestion: `proposal.aborted` check should be done before checking abort window.

Reply:
What's the difference?

Reply:
nothing much, but if proposal is already aborted then why check for abort window? So first check if proposal is already aborted